cpp/src/groupby/hash/groupby.cu

```diff
@@ -258,68 +255,92 @@ void compute_single_pass_aggs(table_view const& keys,
   auto d_values = table_device_view::create(flattened_values);
   rmm::device_vector<aggregation::Kind> d_aggs(aggs);
 
-  bool skip_key_rows_with_nulls = keys_have_nulls and ignore_null_keys;
+  bool skip_key_rows_with_nulls = keys_have_nulls and 
+                                  include_null_keys == include_nulls::NO;
 
-  experimental::detail::grid_1d grid(keys.num_rows(), 256);
   if (skip_key_rows_with_nulls) {
     auto row_bitmask{bitmask_and(keys, rmm::mr::get_default_resource(), stream)};
-    hash::compute_single_pass_aggs<true>
-      <<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>(
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
+      thrust::make_counting_iterator(0), keys.num_rows(),
+      hash::compute_single_pass_aggs<true, Map>{
         map, keys.num_rows(), *d_values, *d_sparse_table, d_aggs.data().get(),
-        static_cast<bitmask_type*>(row_bitmask.data()));
+        static_cast<bitmask_type*>(row_bitmask.data())});
   } else {
-    hash::compute_single_pass_aggs<false>
-      <<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>(
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
+      thrust::make_counting_iterator(0), keys.num_rows(),
+      hash::compute_single_pass_aggs<false, Map>{
         map, keys.num_rows(), *d_values, *d_sparse_table, d_aggs.data().get(),
-        nullptr);
+        nullptr});
   }
 
   // Add results back to sparse_results cache
   auto sparse_result_cols = sparse_table.release();
   for (size_t i = 0; i < aggs.size(); i++) {
-    sparse_results.add_result(col_ids[i],
-                              std::make_unique<aggregation>(aggs[i]),
-                              std::move(sparse_result_cols[i]));
+    sparse_results->add_result(col_ids[i],
+                               std::make_unique<aggregation>(aggs[i]),
+                               std::move(sparse_result_cols[i]));
   }
 }
 
+/**
+ * @brief Computes and returns a device vector containing all populated keys in
+ * `map`. 
+ */
+template <typename Map>
+rmm::device_vector<size_type> extract_populated_keys(Map map,
+  size_type num_keys, cudaStream_t stream = 0)
+{
+  rmm::device_vector<size_type> populated_keys(num_keys);
+
+  auto get_key = [] __device__ (auto const& element) {
+    size_type key, value;
+    thrust::tie(key, value) = element;
+    return key;
+  };
+
+  auto end_it = thrust::copy_if(rmm::exec_policy(stream)->on(stream),
+    thrust::make_transform_iterator(map.data(), get_key),
+    thrust::make_transform_iterator(map.data() + map.capacity(), get_key),
+    populated_keys.begin(),
+    [unused_key = map.get_unused_key()] __device__ (size_type key) {
+      return key != unused_key;
+    });
+
+  populated_keys.resize(end_it - populated_keys.begin());
```

Comment:
I believe with `thrust::device_vector` this is going to incur an additional allocation + memcpy. Seeing as though it's just a temporary used for the gather map, that's probably an unnecessary cost.